compiler/transforms/set_memory_layout.py

```diff
@@ -0,0 +1,99 @@
+from xdsl.dialects import builtin, linalg
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+from compiler.dialects.snax import LayoutCast
+from compiler.dialects.tsl import TiledStridedLayoutAttr
+from compiler.ir.tsl import Stride, TiledStride, TiledStridedLayout
+
+
+class AddMemoryLayout(RewritePattern):
+    """
+    This class represents a rewrite pattern for adding memory layout to a
+    linalg operation. The implementation is very naive. It imposes a specific
+    memory layout on the input and output of the linalg operation dispatched
+    to snax_qgemm by inserting reshuffling ops. In the future, the memory
+    layout will be selected in a more automatic way.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, linalg_op: linalg.Generic, rewriter: PatternRewriter):
+        # check if operation is dispatched via library call
+        if linalg_op.library_call is None:
+            return
+        else:
+            library_call = linalg_op.library_call.data
+
+        # check for library call
+        if library_call == "snax_qgemm":
+            tsl_input_a = TiledStridedLayoutAttr(
+                TiledStridedLayout(
+                    [
+                        TiledStride([Stride(None, None), Stride(8, 8)]),
+                        TiledStride([Stride(256, None), Stride(1, 8)]),
+                    ]
+                )
+            )
+
+            ## tsl b has an offset of 64 to not collide with the banks of a
+            tsl_input_b = TiledStridedLayoutAttr(
+                TiledStridedLayout(
+                    [
+                        TiledStride([Stride(256, None), Stride(1, 8)]),
+                        TiledStride([Stride(None, None), Stride(8, 8)]),
+                    ],
+                    offset=64,
+                )
+            )
+
+            tsl_output = TiledStridedLayoutAttr(
+                TiledStridedLayout(
+                    [
+                        TiledStride([Stride(256, None), Stride(4, 8)]),
+                        TiledStride([Stride(None, None), Stride(32, 8)]),
+                    ]
+                )
+            )
```

Comment:
Would it make sense here to already write this in terms of parametrized banks?

Reply:
I don't think this is worth it, I want to get rid of this code as soon as possible, + this will change a lot for experiments with different data layouts and stuff